### ISSUE TITLE
Annotation priorities

### ISSUE THREAD

**Issue by uliska** [openlilylib/scholarly#24]
Another property annotations should get is a priority (usable for sorting and filtering)

**Comment by jefferyshivers** [comment]
Such as high, normal, low? Where the default is `normal` perhaps?
`priority=high`
So there could be option to sort highest to lowest, or vice versa.
And also option to only print high, or high and normal, etc.

**Comment by uliska** [comment]
Yes, that's what I meant. 

We could think of just using integers instead of keywords, so a user can easily decide how many levels of priority they want.

**Comment by jefferyshivers** [comment]
Should this be a mandatory property?

**Comment by uliska** [comment]
Don't think so. It's nothing every project needs, even less each annotation. I rather suggest defaulting to something corresponding to "regular". For example accept values 1-5 and use 3 as default.

**Comment by jefferyshivers** [comment]
So if I understand, currently they are, by default, sorted by rhythmic location first, then author, then type.

If so, then will we want to sort by priority (of course, only if turned on, and whichever way) first, then sort _that_ by `sort-annotations`?

**Comment by uliska** [comment]
## I wouldn't say so. Default should still be rhythmic lostion IMO.

Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by jefferyshivers** [comment]
In that case, can you clarify for me where priority sorting would take place? Only in latex (or rather, _not_ in LilyPond)?

**Comment by uliska** [comment]
It's a non-trivial question, I think.

It _has_ to be possible to sort/group/filter by priority in LilyPond. The most general use case I can think of is reducing the amount of messages that are irrelevant to the current working stage. There are many users that will _not_ use LaTeX at all (unfortunately ;-) ), and there will be many use cases where you don't want to have all the annotations in the first place, so filtering only in LaTeX doesn't help in these cases.

This _may_ not feel relevant to you now, but I have already worked on a score with more than 1.000 annotations, and it _is_ relevant if you can reduce the console output to, say, 20 messages of type "lilypondIssue" _and_ priority=high (or `> 2` or whatever).

OTOH it seems suboptimal to be forced to recompile the LilyPond score in order to have a different filtering/sorting applied in a LaTeX document. So my impression is we have to implement it on _both_ sides.

But from the perspective of an overall roadmap I would put the LaTeX feature on the list for the final project, with the option to not implement it during GSoC at all. We should start taking care of realizing an appropriate relation of LilyPond and LaTeX development.

**Comment by jefferyshivers** [comment]
Yes I completely agree and think this does make sense to use in LilyPond, but I'm just wondering at what point in the process (in LilyPond) they'll be _sorted_. Since "sorting":

> (usable for sorting and filtering)

... seems to mean something different than filtering. I assumed sorting, then, would mean like high-to-low -OR- low-to-high -OR- not-sorted. And then filtering, separately, would be print-only 4,5 (but obviously with whatever actual syntax), etc.

My suggestion was:

> If so, then will we want to sort by priority (of course, only if turned on, and whichever way) first, then sort that by sort-annotations?

... but if we don't implement sorting that way, then how/where? The answer still isn't clear to me, which is why I wondered if maybe you meant sorting later in LaTeX (which is _not_ what I have been meaning).

**Comment by uliska** [comment]
In LilyPond, filtering is applied first (then annotations aren't added to the list in the first place).

Sorting is the first step in the `annotationProcessor` (`engraver.ily`, line 140ff.).
Selecting and ordering sort operations is done using `(getOption '(scholarly annotate sort-criteria))`, and in `sort.ily` there must be a) a comparison predicate (like `annotation-less-than-by-author?`) and b) a mapping in the `annotation-comparison-predicates` alist.

So it's necessary to add these two items to `sort.ily`, then the user can turn on the sorting by setting the `sort-criteria` option. My suggestion is that by default sorting by priority is simply _not_ added to that option.

Is that clearer now?

**Comment by jefferyshivers** [comment]
Okay yes, thanks for the explanation. That's actually what I expected previously, but 

> If so, then will we want to sort by priority (of course, only if turned on, and whichever way) first, then sort that by sort-annotations?

must have not come across as such. I suggested by "only if turned on" that the default would be _no_ sorting by priority.

**Comment by uliska** [comment]
Yes, and that's what I wanted to indicate by saying that it is not added to the default entries of the `sort-criteria` option.

So yes, I suggest _not_ sorting by criteria by default but only when the user explicitly sets the option.
